client/state/data-layer/README.md

```diff
@@ -0,0 +1,109 @@
+Instead the components should be able to request that they require such data the the middleware will intercept the request to fulfill it accordingly.
+
+## Implementation
+
+The data layer _intercepts_ Redux actions.
+Once it does this the action will be dropped entirely.
+This is to prevent two middlewares which can both supply a given request from fighting with each other or duplicating the request; consequently it allows for prioritization of data-fetching needs by means of ordering how the middleware are arranged in the chain.
+
+Each middleware intercepts given Redux actions and will correspondingly dispatch new follow-up actions to actually handle their requests.
+The functions that compose to form the middleware _can_ and in most cases _will_ closely resemble what was previously written in `redux-thunk` actions.
+
+The middleware should be pieced together from multiple handler functions in a logical way which encourages modularity and isolation of responsibilities.
+The data layer provides `mergeHandlers()` as a utility to support this.
+**Each file should export an object whose keys are Redux action types to intercept and whose values are lists of functions to invoke on those actions.**
+The list of handlers will be called in sequence in the order given in the list itself and in the order in which the different handlers are merged.
+Even when there is only one handler the value of the exported key should be a list - a singleton list in this case.
+
+```js
+const requestSplines = ( { dispatch }, action ) => {
+	…
+};
+
+const invalidateExisting = ( { dispatch } ) => {
+	…
+};
+
+const requestReticulations = ( store, { splineSet } ) =>
+	requestSplines( store, { type: REQUEST_SPLINES, setId: splineSet } );
+
+export default {
+	[ REQUEST_RETICULATIONS ]: [ requestReticulations ],
+	[ REQUEST_SPLINES ]: [ invalidateExisting, requestSplines ],
+}
+
+// middlware.js
+const handlers = mergeHandlers(
+	postsEndpoint,
+	sitesEndpoint,
+	splineEndpoint,
+);
+```
+
+Note that **these middlewares are not restricted to making HTTP requests**.
+A middleware might end up communicating over a `WebSocket`, making `HTTP` requests, reading from a local JSON file, or doing any other operation permitted in JavaScript.
+There could be special middleware running on the desktop platform which invokes code which is only possible in the desktop environment, for example.
+
+**These also need not handle every possible action type**.
+This system is incremental and can replace pieces of the data synchronization system.
+In some cases it might be appropriate for a certain middleware to only handle one type of data; this is fine.
+
+## File Structure
+
+Since this middleware mirrors the WordPres.com API its files should also mirror its structure. For example, the file which handles requests to the `/sites/[ siteID ]/posts` endpoint should live at `state/data-layer/wpcom/sites/posts`.
```

Comment:
It's unclear from this paragraph whether it should live at `posts.js` or `posts/index.js`. Considering your provided example `plans.js`, where would we add a theoretical future `/plans/free` endpoint? Should we instead have the base endpoint live at `plans/index.js` ?